src/S3/S3Transfer/S3TransferManager.php

```diff
@@ -0,0 +1,803 @@
+<?php
+
+namespace Aws\S3\S3Transfer;
+
+use Aws\S3\S3Client;
+use Aws\S3\S3ClientInterface;
+use Aws\S3\S3Transfer\Exceptions\S3TransferException;
+use Aws\S3\S3Transfer\Models\DownloadDirectoryRequest;
+use Aws\S3\S3Transfer\Models\DownloadDirectoryResponse;
+use Aws\S3\S3Transfer\Models\DownloadFileRequest;
+use Aws\S3\S3Transfer\Models\DownloadRequest;
+use Aws\S3\S3Transfer\Models\S3TransferManagerConfig;
+use Aws\S3\S3Transfer\Models\UploadDirectoryRequest;
+use Aws\S3\S3Transfer\Models\UploadDirectoryResponse;
+use Aws\S3\S3Transfer\Models\UploadRequest;
+use Aws\S3\S3Transfer\Models\UploadResult;
+use Aws\S3\S3Transfer\Progress\MultiProgressTracker;
+use Aws\S3\S3Transfer\Progress\SingleProgressTracker;
+use Aws\S3\S3Transfer\Progress\TransferListener;
+use Aws\S3\S3Transfer\Progress\TransferListenerNotifier;
+use Aws\S3\S3Transfer\Progress\TransferProgressSnapshot;
+use Aws\S3\S3Transfer\Utils\DownloadHandler;
+use FilesystemIterator;
+use GuzzleHttp\Promise\Each;
+use GuzzleHttp\Promise\PromiseInterface;
+use InvalidArgumentException;
+use Psr\Http\Message\StreamInterface;
+use RecursiveDirectoryIterator;
+use RecursiveIteratorIterator;
+use function Aws\filter;
+use function Aws\map;
+
+class S3TransferManager
+{
+    /** @var S3Client  */
+    private S3ClientInterface $s3Client;
+
+    /** @var S3TransferManagerConfig  */
+    private S3TransferManagerConfig $config;
+
+    /**
+     * @param S3ClientInterface | null $s3Client If provided as null then,
+     * a default client will be created where its region will be the one
+     * resolved from either the default from the config or the provided.
+     * @param array|S3TransferManagerConfig|null $config
+     */
+    public function __construct(
+        ?S3ClientInterface $s3Client = null,
+        array|S3TransferManagerConfig|null $config = null
```

Comment:
question: should we make this a required parameter? what's the use case for accepting a null value if we need things like `Bucket` and `Key`?

Reply:
I don't understand the question. Is it in the right file position?

Reply:
probably should have looked at the `S3TransferManagerConfig` before I assumed it had required parameters.  Disregard